### PR TITLE
Average training and validation losses using torch.distributed for monitoring trends

### PR DESCRIPTION
**Pull request by bumi001**
1. Added code for deterministic computation  
   - helps in reproducing results
   - expects torch and cuda versions to be the same
2. Added code to calculate average training loss over the entire training dataset across all gpus
   - uses torch.distributed.reduce
   - printed on a per epoch basis
   - helps monitor training trend.
3. Added code to calculate average validation loss over the entire validation dataset across all gpus
   - uses torch.distributed.reduce
   - printed whenever it is calculated
   - helps monitor validation trend.
4. Added code to save the history of average training loss and average validation loss, as needed, as a function of epoch.

### CODE REVIEW

File: gns/train.py [hunk 1]
```diff
@@ -464,6 +636,9 @@ def main(_):
 
   myflags = reading_utils.flags_to_dict(FLAGS)
 
+  myflags["validation_interval"] = 3
+  myflags["nsave_steps"] = 5
+
```
Comment (kks32) [inline gns/train.py:641]:
```suggestion

```

File: gns/train.py [hunk 2]
```diff
@@ -21,6 +21,17 @@
...
+seed=42
+torch.manual_seed(seed)
+torch.cuda.manual_seed(seed)
+torch.backends.cudnn.deterministic = True
+random.seed(seed)
+np.random.seed(seed)
+torch.use_deterministic_algorithms(True)
+# do this on a bash shell
+# export CUBLAS_WORKSPACE_CONFIG=:4096:8
+
```
Comment (kks32) [inline gns/train.py:34]:
```suggestion

```

File: gns/train.py [hunk 3]
```diff
@@ -358,26 +431,62 @@ def train(rank, flags, world_size, device):
...
-            else:
-              simulator.module.save(flags["model_path"] + 'model-'+str(step)+'.pt')
-            train_state = dict(optimizer_state=optimizer.state_dict(),
-                               global_train_state={"step": step},
-                               loss=loss.item())
-            torch.save(train_state, f'{flags["model_path"]}train_state-{step}.pt')
-
         # Complete training
-        if (step >= flags["ntraining_steps"]):
+        # if (step >= flags["ntraining_steps"]):
```
Comment (kks32) [inline gns/train.py:435]:
```suggestion

```

File: gns/train.py [hunk 4]
```diff
@@ -358,26 +431,62 @@ def train(rank, flags, world_size, device):
...
+        if (epoch >= flags["ntraining_steps"]):
           not_reached_nsteps = False
           break
 
+        # if epoch == 0:
+        #   print(f"step: {step}, loss: {loss}, total_loss: {total_loss}")
+
```
Comment (kks32) [inline gns/train.py:442]:
```suggestion

```

**Review by kks32 — CHANGES_REQUESTED** [review]
Could you please remove the hardcoded steps and random seed from this PR

File: gns/train.py (hunk 2 above)
Comment (bumi001) [reply gns/train.py]: I suggest commenting these out, just in case someone wants reproducible outputs.
